org.eclipse.jdt.launching/launching/org/eclipse/jdt/internal/launching/DetectVMInstallationsJob.java

```diff
@@ -0,0 +1,154 @@
+import org.eclipse.jdt.launching.VMStandin;
+
+/**
+ * Lookup for VMs installed in standard or usual locations; and add the existing ones that
+ * are not yet known by JDT to the VM registry (usually visible in the "Installed JREs"
+ * preference page)
+ */
+class DetectVMInstallationsJob extends Job {
+
+	private static final Object FAMILY = DetectVMInstallationsJob.class;
+
+	private final StandardVMType standardType;
+	
+	DetectVMInstallationsJob() {
+		super(LaunchingMessages.lookupInstalledJVMs);
+		this.standardType = (StandardVMType)JavaRuntime.getVMInstallType(StandardVMType.ID_STANDARD_VM_TYPE);
+	}
+
+	@Override
+	protected IStatus run(IProgressMonitor monitor) {
+		Collection<File> candidates = computeCandidateVMs();
+		if (monitor.isCanceled()) {
+			return Status.CANCEL_STATUS;
+		}
+		Set<File> knownVMs = knownVMs();
+		candidates.removeIf(knownVMs::contains);
+		monitor.beginTask(LaunchingMessages.lookupInstalledJVMs, candidates.size());
+		for (File f : candidates) {
+			if (monitor.isCanceled()) {
+				return Status.CANCEL_STATUS;
+			}
+			SubMonitor subMon = SubMonitor.convert(monitor, f.getAbsolutePath(), 1);
+			VMStandin workingCopy = new VMStandin(standardType, f.getAbsolutePath());
+			workingCopy.setInstallLocation(f);
+			String name = f.getName();
+			int i = 1;
+			while (isDuplicateName(name)) {
+				name = f.getName() + '(' + i++ + ')';
+			}
+			workingCopy.setName(name);
+			IVMInstall install = workingCopy.convertToRealVM();
+			if (!(install instanceof IVMInstall2 vm && vm.getJavaVersion() != null)) {
+				// worksaround: such VMs may cause issue later
+				// https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/248
+				standardType.disposeVMInstall(install.getId());
+			}
+			subMon.done();
+		}
+		return Status.OK_STATUS;
+	}
+
+	private boolean isDuplicateName(String name) {
+		return Stream.of(JavaRuntime.getVMInstallTypes()) //
+			.flatMap(vmType -> Arrays.stream(vmType.getVMInstalls())) //
+			.map(IVMInstall::getName) //
+			.anyMatch(name::equals);
+	}
+
+	private Collection<File> computeCandidateVMs() {
+		// parent directories containing a collection of VM installations
+		Collection<File> rootDirectories = new HashSet<>();
+		if (!Platform.OS_WIN32.equals(Platform.getOS())) {
+			rootDirectories.add(new File("/usr/lib/jvm")); //$NON-NLS-1$
+		}
+		if (Platform.OS_MACOSX.equals(Platform.getOS())) {
+			rootDirectories.add(new File("/Library/Java/JavaVirtualMachines")); //$NON-NLS-1$
+		}
+		rootDirectories.add(new File(System.getProperty("user.home"), ".sdkman/candidates/java")); //$NON-NLS-1$ //$NON-NLS-2$
+
+		Set<File> directories = rootDirectories.stream().filter(File::isDirectory)
+			.map(dir -> dir.listFiles(File::isDirectory))
+			.flatMap(Arrays::stream)
+			.filter(Objects::nonNull)
+			.collect(Collectors.toSet());
+
+		// particular VM installations
+		String javaHome = System.getenv("JAVA_HOME"); //$NON-NLS-1$
+		if (javaHome != null) {
+			directories.add(new File(javaHome));
```

Comment:
Directories may be unmodifiable. See [Collectors.toSet()](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/stream/Collectors.html#toSet())
> There are no guarantees on the type, mutability, serializability, or thread-safety of the Set returned; if more control over the returned Set is required, use [toCollection(Supplier)](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/stream/Collectors.html#toCollection(java.util.function.Supplier)).